### ISSUE TITLE
🐛 Bug: `isExpression` probably doesn't match all expressions

### ISSUE THREAD

**Issue by RebeccaStevens** [JoshuaKGoldberg/ts-api-utils#31]
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/ts-api-utils/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

N/A

### Actual

N/A

### Additional Info

Unlike other general types that are a union of more precise types (e.g. `type BooleanLiteral = TrueLiteral | FalseLiteral`), `Expression` is just an interface; it doesn't even declare any non-inherited properties/methods (other than a brand).

The way the function currently works is by checking if the given node is one of many known expression types, but this list could easily become (and probably is) incomplete. Doing a search through TypeScript's type declaration file, there are several types with the "Expression" suffix that aren't included in the current list of types that the function checks against.

~~I think the best way to approach the implementation is probably just to test if the node has the expression brand (`typeof (node as ts.Expression)._expressionBrand !== undefined`).~~
Edit: Branding doesn't seem to be available at runtime.

**Comment by RebeccaStevens** [comment]
TypeScript has its own `isExpression` and `isExpressionNode` functions. I'd say it best to probably just hook straight into using these instead of trying to recreate them.

The `isExpressionNode` one seems to be closes to how tsutils' `isExpression` function works.

**Comment by JoshuaKGoldberg** [comment]
> Branding doesn't seem to be available at runtime

Yeah, it's the TypeScript convention for branded types. #202 and #5855 have interesting discussion & links around the topic.

> TypeScript has its own isExpression and isExpressionNode functions

They're private & not in the types though 😞. The closest public one is `ts.isExpressionStatement`. I'm very hesitant to take a dependency on those private APIs.

Once #3 is done, I'd be interested in filing an issue on TypeScript asking which internal functions of theirs they'd be willing to make public.

**Comment by RebeccaStevens** [comment]
> I'd be interested in filing an issue on TypeScript asking which internal functions of theirs they'd be willing to make public.

I remember many years back when I first started using the TypeScript API, I opened an issue along those lines for a function I wanted to use that was private. Essentially, I was just told to use `tsutils` (which I hadn't heard of at the time).
As `tsutils` re-defined a bunch of these type guards I was using, I decided to just not import any type guards directly from `typescript` and exclusively use the `tsutils` ones.

I'm not entirely sure why TypeScript makes some of their type guards private in the first place. Hopefully we can convince them to expose a few more of them. (In #33 `isExpression` and `isLeftHandSideExpression` are needed for some of the union and compound type guards and I don't think there's any way around that)
